### PR TITLE
Game Test API

### PR DESCRIPTION
**Pull request by LambdAurora**
One of the most important addition for stronger testing!

This API is, well, the game test API, which allows mods to use the Minecraft test automation framework ([keynote about it by Henrik Kniberg](https://www.youtube.com/watch?v=TNkPE6NTNHQ)).

This API introduces a new entrypoint `quilt:game_test` of classes implementing `QuiltGameTest`. Any method annotated with `GameTest` in that class will be registered as a test, and you can implement a method from `QuiltGameTest` to programmatically register more test classes.

The API is partially a port from Fabric API's game test API, but extended to be more comfortable to use, and more powerful!

As you can see there's also some game tests implemented for Block Content Registries to show how it can be used (and its benefits).

### CODE REVIEW

**Review by Leo40Git — APPROVED** [review]

**Review by OroArmor — APPROVED** [review]

**Review by Mysterious-Dev — APPROVED** [review]

File: build.gradle [hunk 1]
```diff
@@ -165,6 +165,16 @@ loom {
 			configName = "Auto test server"
 			property("quilt.auto_test")
 		}
+
+		// Game test server, a server is executes tests using Minecraft's test framework.
```
Comment (EnnuiL) [inline build.gradle]:
```suggestion
		// Game test server, a server that executes tests using Minecraft's test framework.
```

File: gradle.properties [hunk 2]
```diff
@@ -1 +1,3 @@
-org.gradle.jvmargs=-Xmx2048m
+# Bump this if you're encountering OutOfMemory errors,
+# though needing more might be a sign that something is wrong in the build tool.
+org.gradle.jvmargs=-Xmx3072m
```
Comment (EnnuiL) [inline gradle.properties]: oh no it's time to download more ram
(although i think the 3 GBs limit should be fine on a 8 GB PC)
(you can ignore this comment)

File: library/block/block_content_registry/src/testmod/java/org/quiltmc/qsl/block/test/BlockContentRegistryTest.java [hunk 3]
```diff
@@ -86,4 +126,44 @@ private <T> void assertValues(Block block, RegistryEntryAttachment<Block, T> att
...
+					this.entries.forEach(entry ->
+							context.checkBlockState(entry.pos(), state -> state.equals(entry.targetState()),
+									() -> "Could not find state " + entry.targetState()
+							)
+					)
+			);
+		}
+
+		record Entry(BlockPos pos, BlockState baseState, BlockState targetState) {
+		}
```
Comment (EnnuiL) [inline library/block/block_content_registry/src/testmod/java/org/quiltmc/qsl/block/test/BlockContentRegistryTest.java]:
```suggestion
		record Entry(BlockPos pos, BlockState baseState, BlockState targetState) {}
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ * <p>
+ * <h3>Get Started</h3>
+ * Writing tests for your mod is relatively simple: create a new class that extends {@link org.quiltmc.qsl.game_test.api.QuiltGameTest}, and add it
+ * as an entrypoint of key {@value org.quiltmc.qsl.game_test.api.QuiltGameTest#ENTRYPOINT_KEY}.
+ * <p>
+ * Now you can add public methods (may or may not be {@code static}) to the class,
+ * they take as argument {@link org.quiltmc.qsl.game_test.api.QuiltTestContext} which is an extension of {@link net.minecraft.test.TestContext},
+ * and annotating them with {@link net.minecraft.test.GameTest} will make them a test to run.
+ * <p>
+ * One of the special parts of game tests is they have a structure associated to them, as the goal is to test inside a game world with running
```
Comment (EnnuiL) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: hm, maybe this would be better?
```suggestion
 * One of the special parts of game tests is that they have a structure associated to them, as the goal is to test inside a game world with running
```

**Review by EnnuiL — CHANGES_REQUESTED** [review]
oh wow, this PR looks great!
I only have few nitpicks

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ * <h3>Custom Test Method Invocation</h3>
+ * Test methods are gathered using the {@link net.minecraft.test.GameTest} annotation, but for methods which the class implements
+ * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest} you can override the method
+ * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest#invokeTestMethod(QuiltTestContext, org.quiltmc.qsl.game_test.api.TestMethod)}
+ * which override entirely how the test is invoked, the default invocation is done using
+ * {@link org.quiltmc.qsl.game_test.api.TestMethod#invoke(java.lang.Object, java.lang.Object...)} with the instance of the class (if the method isn't
+ * {@code static}), and as first parameter the test context.
+ */
+
+package org.quiltmc.qsl.game_test.api;
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: ew! is snake_case really the convention here?

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
+/**
+ * <h2>The Game Test API.</h2>
+ *
+ * <p>
+ * <h3>What is the Game Test API?</h3>
+ * The Game Test API is the test automation framework used by Minecraft as most test frameworks are not adapted to a game that's so dynamic.
+ * Unit testing presents very quickly its issues: you will quickly find out that testing while in a tick loop is very difficult.
+ * Thus Mojang made their own testing framework, watching the <a href="https://www.youtube.com/watch?v=TNkPE6NTNHQ">keynote by Henrik Kniberg</a>
+ * is highly recommended to grasp the why, and how it works.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * The Game Test API is the custom automated test framework used by Minecraft.
 * Minecraft uses a custom framework because most pre-existing ones do not adapt well to such a dynamic environment as Minecraft. 
 * Thus Mojang made their own testing framework. Watching the <a href="https://www.youtube.com/watch?v=TNkPE6NTNHQ">keynote by Henrik Kniberg</a>
 * is highly recommended to grasp how it works and why it was designed that way.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * <p>
+ * The Game Test API exposed by this module is a bit different, the framework has been adapted to work better with mods, with some extensions.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: Different to what? Here's a revision based on what I think you meant:

```suggestion
 * The Game Test API exposed by this module is a bit different to what Vanilla provides. The framework has been adapted to work better with mods, with some extensions.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ * <h3>Get Started</h3>
+ * Writing tests for your mod is relatively simple: create a new class that extends {@link org.quiltmc.qsl.game_test.api.QuiltGameTest}, and add it
+ * as an entrypoint of key {@value org.quiltmc.qsl.game_test.api.QuiltGameTest#ENTRYPOINT_KEY}.
+ * <p>
+ * Now you can add public methods (may or may not be {@code static}) to the class,
+ * they take as argument {@link org.quiltmc.qsl.game_test.api.QuiltTestContext} which is an extension of {@link net.minecraft.test.TestContext},
+ * and annotating them with {@link net.minecraft.test.GameTest} will make them a test to run.
+ * <p>
+ * One of the special parts of game tests is that they have a structure associated to them, as the goal is to test inside a game world with running
+ * game logic. Structures are located in the mod resources as follow {@code data/<id namespace>/game_test/structures/<id path>.snbt}.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * game logic. Structures are located in the mod resources as follows: {@code data/<id namespace>/game_test/structures/<id path>.snbt}.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+ * If we take the provided empty structure {@link org.quiltmc.qsl.game_test.api.QuiltGameTest#EMPTY_STRUCTURE {@code quilt:empty}}, it is located in
+ * {@code data/quilt/game_test/structures/empty.snbt}.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * For example, the provided empty structure {@link org.quiltmc.qsl.game_test.api.QuiltGameTest#EMPTY_STRUCTURE {@code quilt:empty}} is located at
 * {@code data/quilt/game_test/structures/empty.snbt}.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+ * <p>
+ * When running the game with this module a new command will be available: {@code /test}. It allows to create new test structures,
+ * export test structures, run a specific test, or run every tests.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * When running the game with this module a new command will be available: {@code /test}. It allows you to create new test structures,
 * export test structures, run a specific test, or run every test.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+ * <p>
+ * To run the dedicated server in game test mode (as-in runs, execute the tests, then shutdown), you can pass the argument
+ * {@code -Dquilt.game_test=true} to the JVM, you also can pass the argument {@code -Dquilt.game_test.report_file=<path>} to store the test results
+ * into the file at the given path.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * To run the dedicated server in game test mode (which runs the server, executes the tests, then shuts down), you can pass the argument
 * {@code -Dquilt.game_test=true} to the JVM. You also can pass the argument {@code -Dquilt.game_test.report_file=<path>} to store the test results
 * into the file at the given path.
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+ * <p>
+ * <h4>Example</h4>
+ * <pre><code>
+ * public class ExampleTestSuite implements QuiltGameTest {
+ * 	&#64;GameTest(structureName = QuiltGameTest.EMPTY_STRUCTURE)
+ * 	public void noStructure(QuiltTestContext context) {
+ * 		context.setBlockState(0, 2, 0, Blocks.DIAMOND_BLOCK);
+ *
+ * 		context.addInstantFinalTask(() ->
+ * 				context.checkBlock(new BlockPos(0, 2, 0), (block) -> block == Blocks.DIAMOND_BLOCK, "Expect block to be diamond")
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * 				context.checkBlock(new BlockPos(0, 2, 0), (block) -> block == Blocks.DIAMOND_BLOCK, "Expected block to be diamond")
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+ * 		);
+ * 	}
+ * }
+ * </code></pre>
+ * <p>
+ * <h4>Gradle Setup Example</h4>
+ * Here's an example of Gradle setup to handle tests in a separate sourceSet, with a special task to run the tests:
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * Here's an example of a Gradle setup to handle tests in a separate sourceSet, with a special task to run the tests:
```

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+ * then in every {@link net.minecraft.test.GameTest} of the class you only need what's following after the prefix for the structure name.
+ *
+ * <p>
+ * <h3>Custom Test Method Invocation</h3>
+ * Test methods are gathered using the {@link net.minecraft.test.GameTest} annotation, but for methods which the class implements
+ * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest} you can override the method
+ * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest#invokeTestMethod(QuiltTestContext, org.quiltmc.qsl.game_test.api.TestMethod)}
+ * which override entirely how the test is invoked, the default invocation is done using
+ * {@link org.quiltmc.qsl.game_test.api.TestMethod#invoke(java.lang.Object, java.lang.Object...)} with the instance of the class (if the method isn't
+ * {@code static}), and as first parameter the test context.
```
Comment (TheGlitch76) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]:
```suggestion
 * Test methods are gathered using the {@link net.minecraft.test.GameTest} annotation, but if the class implements
 * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest} you can override the method
 * {@link org.quiltmc.qsl.game_test.api.QuiltGameTest#invokeTestMethod(QuiltTestContext, org.quiltmc.qsl.game_test.api.TestMethod)}
 * which entirely overrides how the test is invoked; the default invocation is done using
 * {@link org.quiltmc.qsl.game_test.api.TestMethod#invoke(java.lang.Object, java.lang.Object...)} with the instance of the class (if the method isn't
 * {@code static}), and the test context as the first parameter.
```

**Review by TheGlitch76 — CHANGES_REQUESTED** [review]
The documentation in the package-info is really nice! I provided some really picky corrections, but the structure is really really good.

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java (hunk 5 above)
Comment (LambdAurora) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: It's not really, and I would push against in most cases

but having a game package just is weird? and I'm worried that a bare test package would make people think they imported something internal?

Thoughts to have

Comment (EnnuiL) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: Wait hmm, isn't GameTest a single word though? Maybe gametest would make sense then

Comment (LambdAurora) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: The exact thing I tried to avoid because I absolutely despite it as a single word and the actual data directory in data pack for structures use `game_test` :blob_waaaaa:

Comment (EnnuiL) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: wait oh no, it's a single word *and* two different ones at the same time? dear god mojang

Comment (LambdAurora) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: Note: we are in control of the directory because Vanilla does not define one clearly

and like, it's capitalized `GameTest` not `Gametest`.

Comment (williambl) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: Personally I don't mind using `game_test`. Having `gametest` when we also call it 'Game Test' and 'GameTest' feels wrong.

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/mixin/server/MainMixin.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+import org.quiltmc.loader.api.minecraft.DedicatedServerOnly;
+import org.quiltmc.qsl.game_test.impl.QuiltGameTestImpl;
+
+@DedicatedServerOnly
+@Mixin(Main.class)
+public class MainMixin {
+	@Redirect(method = "main", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/dedicated/EulaReader;isEulaAgreedTo()Z"))
+	private static boolean isEulaAgreedTo(EulaReader reader) {
+		return QuiltGameTestImpl.ENABLED || reader.isEulaAgreedTo();
```
Comment (williambl) [inline library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/mixin/server/MainMixin.java]: Do we need/have some kind of "By running these tests you agree to the Minecraft EULA" message if we're doing this?

**Review by williambl — COMMENTED** [review]
Looks good! I'll give it a proper look/test in my IDE later :D

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/mixin/server/MainMixin.java (hunk 15 above)
Comment (LambdAurora) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/mixin/server/MainMixin.java]: While I'm not entirely convinced by the need of such message (especially considering the very nature of modding), I guess it doesn't hurt to put such message.

File: library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java (hunk 5 above)
Comment (TheGlitch76) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: Common convention is to mash everything into one word (mypackage, hackmd, stackoverflow, the list goes on). I don't really understand why we are being different here.

Comment (LambdAurora) [reply library/core/game_test/src/main/java/org/quiltmc/qsl/game_test/api/package-info.java]: I have thought of another package that would fit better being `testing.game`
In retrospect that new package name would fit better and would allow other testing modules in the future and not limiting ourselves to just "game"
Thus increasing package discovery yada yada

tl;dr to me `testing.game` makes more sense now

**Comment by EnnuiL** [comment]
Hm, as a precaution, we should probably consider merging this [GameTest mappings PR](https://github.com/QuiltMC/quilt-mappings/pull/324) before doing any further push for release

**Comment by LambdAurora** [comment]
> Hm, as a precaution, we should probably consider merging this [GameTest mappings PR](https://github.com/QuiltMC/quilt-mappings/pull/324) before doing any further push for release

that mappings PR doesn't target the version this PR targets anymore :)

**Comment by EnnuiL** [comment]
> > Hm, as a precaution, we should probably consider merging this [GameTest mappings PR](https://github.com/QuiltMC/quilt-mappings/pull/324) before doing any further push for release
> 
> that mappings PR doesn't target the version this PR targets anymore :)

Not if we downgrade it back!

**Comment by EnnuiL** [comment]
> > > Hm, as a precaution, we should probably consider merging this [GameTest mappings PR](https://github.com/QuiltMC/quilt-mappings/pull/324) before doing any further push for release
> > 
> > 
> > that mappings PR doesn't target the version this PR targets anymore :)
> 
> Not if we downgrade it back!

And with some poking? My dreams became true! :P

But yeah, the PR has been merged on the 22w03a branch and backported to the 1.19.3 one on QM 1.19.3-build.18

**Review by EnnuiL — APPROVED** [review]
The only change that I'd make is adding the "You agree to the EULA by running these tests" message, but yeah, it can be done after merge

**Review by EnnuiL — APPROVED** [review]

**Review by OroArmor — APPROVED** [review]
